services/datamanager/model/model.go

```diff
@@ -0,0 +1,324 @@
+		client:     client,
+		logger:     logger,
+		cancelFunc: cancelFunc,
+		partID:     partID,
+		httpClient: httpClient,
+	}
+	return &ret, nil
+}
+
+func (m *modelManager) deploy(ctx context.Context, req *v1.DeployRequest) (*v1.DeployResponse, error) {
+	resp, err := m.client.Deploy(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+// Close all resources (goroutines) associated with modelManger.
+func (m *modelManager) Close() {
+	m.cancelFunc()
+	if m.conn != nil {
+		if err := m.conn.Close(); err != nil {
+			m.logger.Errorw("error closing model deploy server connection", "error", err)
+		}
+	}
+}
+
+// DownloadModels handles deploying models into their specified destination.
+func (m *modelManager) DownloadModels(cfg *config.Config, modelsToDeploy []*Model, errorChannel chan error) chan error {
+	modelsToDownload, err := getModelsToDownload(modelsToDeploy)
+	if err != nil {
+		errorChannel <- err
+		return errorChannel
+	}
+	// TODO: DATA-295, delete models in file system that are no longer in the config. If we have no models to download, exit.
+	if len(modelsToDownload) == 0 {
+		return nil
+	}
+	// Stop download of previous models if we're trying to download new ones.
+	if m.cancelFunc != nil {
+		m.cancelFunc()
+	}
+
+	cancelCtx, cancelFn := context.WithCancel(context.Background())
+	m.cancelFunc = cancelFn
+	for _, model := range modelsToDownload {
+		deployRequest := &v1.DeployRequest{
+			Metadata: &v1.DeployMetadata{
+				ModelName: model.Name,
+			},
+		}
+		deployResp, err := m.deploy(cancelCtx, deployRequest)
+		if err != nil {
+			m.logger.Error(err)
+			errorChannel <- err
+		}
+		url := deployResp.Message
+		filePath := filepath.Join(model.Destination, model.Name)
+		err = downloadFile(cancelCtx, m.httpClient, filePath, url, m.logger)
+		if err != nil {
+			m.logger.Error(err)
+			errorChannel <- err
+		}
+		// A download from a GCS signed URL only returns one file.
+		modelFileToUnzipPath := filepath.Join(model.Destination, model.Name+zipExtension)
+		if err = unzipSource(modelFileToUnzipPath, model.Destination); err != nil {
+			m.logger.Error(err)
+			errorChannel <- err
+		}
+	}
+	close(errorChannel)
+	return errorChannel
+}
+
+// getModelsToDownload fetches the models that need to be downloaded according to the
+// provided config.
+func getModelsToDownload(models []*Model) ([]*Model, error) {
+	// TODO: DATA-405, if the user specifies one destination to deploy their models into
+	// this function will not work as expected.
+	modelsToDownload := make([]*Model, 0)
+	for _, model := range models {
+		if model.Destination == "" {
+			// Set the model destination to default if it's not specified in the config.
+			model.Destination = filepath.Join(ViamModelDotDir, model.Name)
+		}
+		_, err := os.Stat(model.Destination)
+		switch {
+		case errors.Is(err, os.ErrNotExist):
+			// we know we have never attempted to deploy the model
+			modelsToDownload = append(modelsToDownload, model)
```

Comment:
supernit: can add `modelsToDownload` right after trying to make the model.Destination directory instead of right before, since no need to append if erroring out

Reply:
agreed.